panic when stopping vm

# Required information

 * Distribution: Ubuntu
 * Distribution version: 20.10
 * The output of "lxc info" or if that fails:
   * Kernel version: 5.8.0-33-generic #36-Ubuntu SMP Wed Dec 9 09:14:40 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
   * LXC version: 4.0.0-devel
   * LXD version: 4.9
   * Storage backend in use: btrfs

# Issue description

Panic in `qmp` when running `lxc stop <vm> --force`

# Steps to reproduce

Not completely sure.

# Information to attach

 - [ ] Any relevant kernel output (`dmesg`)
 - [ ] Container log (`lxc info NAME --show-log`)
 - [ ] Container configuration (`lxc config show NAME --expanded`)
 - [ ] Main daemon log (at /var/log/lxd/lxd.log or /var/snap/lxd/common/lxd/logs/lxd.log)
 - [ ] Output of the client with --debug
 - [ ] Output of the daemon with --debug (alternatively output of `lxc monitor` while reproducing the issue)

```
DBUG[01-04|12:49:39] Instance stopped                         instanceType=virtual-machine instance=f2-vm project=default
DBUG[01-04|12:49:39] Stopping device                          instance=f2-vm project=default instanceType=virtual-machine device=root type=disk
DBUG[01-04|12:49:39] Stopping device                          instance=f2-vm project=default instanceType=virtual-machine device=eth0 type=nic
DBUG[01-04|12:49:40] Clearing instance firewall static filters project=default instance=f2-vm parent=lxdbr0 dev=eth0 host_name=tapfc1ce2e8 hwaddr=00:16:3e:42:74:4d ipv4=0.0.0.0 ipv6=::
DBUG[01-04|12:49:40] Clearing instance firewall dynamic filters ipv4=<nil> ipv6=<nil> project=default instance=f2-vm parent=lxdbr0 dev=eth0 host_name=tapfc1ce2e8 hwaddr=00:16:3e:42:74:4d
DBUG[01-04|12:49:40] UnmountInstance started                  driver=btrfs pool=btrfs1 instance=f2-vm project=default
DBUG[01-04|12:49:40] UnmountInstance finished                 driver=btrfs pool=btrfs1 instance=f2-vm project=default
panic: close of closed channel

goroutine 13502 [running]:
github.com/lxc/lxd/lxd/instance/drivers/qmp.(*Monitor).Disconnect(0xc000216940)
        /home/brauner/src/go/src/github.com/lxc/lxd/lxd/instance/drivers/qmp/monitor.go:176 +0x103
github.com/lxc/lxd/lxd/instance/drivers/qmp.(*Monitor).run.func1()
        /home/brauner/src/go/src/github.com/lxc/lxd/lxd/instance/drivers/qmp/monitor.go:86 +0x2b5
created by github.com/lxc/lxd/lxd/instance/drivers/qmp.(*Monitor).run.func2
        /home/brauner/src/go/src/github.com/lxc/lxd/lxd/instance/drivers/qmp/monitor.go:149 +0x208
```

--- a/lxd/instance/drivers/qmp/monitor.go
+++ b/lxd/instance/drivers/qmp/monitor.go
@@ -171,16 +171,17 @@ func (m *Monitor) Wait() (chan struct{}, error) {
 
 // Disconnect forces a disconnection from QEMU.
 func (m *Monitor) Disconnect() {
+	monitorsLock.Lock()
+	defer monitorsLock.Unlock()
+
 	// Stop all go routines and disconnect from socket.
 	if !m.disconnected {
 		close(m.chDisconnect)
+		m.disconnected = true
+		m.qmp.Disconnect()
 	}
-	m.disconnected = true
-	m.qmp.Disconnect()
 
 	// Remove from the map.
-	monitorsLock.Lock()
-	defer monitorsLock.Unlock()
 	delete(monitors, m.path)
 }
 

